packages/venia-concept/src/ServiceWorker/Utilities/messageHandler.js

```diff
@@ -0,0 +1,125 @@
+};
+
+/**
+ * unRegisterMessageHandler is a util that will un register
+ * a handler for a given message type in the handlers object.
+ *
+ * The handler function provided here should be the exact function
+ * that was used to register in the first place.
+ *
+ * @param {string} type type of the message handler to unregister
+ * @param {function} handler reference of the handler to unregister
+ *
+ * @returns {void}
+ */
+export const unRegisterMessageHandler = (type, handler) => {
+    if (handlers[type]) {
+        handlers[type] = handlers[type].filter(
+            handlerfn => handler !== handlerfn
+        );
+    }
+};
+
+/**
+ *
+ * handleMessageFromClient handles messages from the client. It matches
+ * the message with the handler in the handlers object and delegates
+ * the message payload along with the MessageEvent to the handler.
+ *
+ * @param {string} type type of the message from the client
+ * @param {object} payload payload of the message from the client
+ * @param {MessageEvent} event MessageEvent of the message from the client
+ *
+ * @returns {void}
+ */
+export const handleMessageFromClient = (type, payload, event) => {
+    const handlerList = handlers[type];
+    if (handlerList) {
+        handlerList.forEach(handler => {
+            handler(payload, event);
+        });
+    }
+};
+
+/**
+ *
+ * sendMessageToClient sends a message to the given client. It constructs the
+ * message with the type and payload provided as arguments and wraps them in a
+ * promise that will be returned to the caller. If the client responds to the
+ * message, the promise will resolve with that response.
+ *
+ * @param {Client} client client to send the message
+ * @param {string} type type of the message to be sent to the client
+ * @param {object} payload payload of the message to be sent to the client
+ *
+ * @returns {Promise} promise that will resolve to the reply from the client
+ */
+export const sendMessageToClient = (client, type, payload) =>
+    new Promise((resolve, reject) => {
+        const channel = new MessageChannel();
+
+        channel.port1.onmessage = event => {
+            if (event.data.error) {
+                reject(event.data.error);
+            } else {
+                resolve(event.data);
+            }
+        };
+
+        client.postMessage({ type, payload }, [channel.port2]);
+    });
+
+/**
+ *
+ * sendMessageToWindow sends a message to the window. It constructs the
+ * message with the type and payload provided as arguments and wraps them in a
+ * promise that will be returned to the caller. If the window responds to the
+ * message, the promise will resolve with that response.
+ *
+ * @param {string} type type of the message to be sent to the window
+ * @param {object} payload payload of the message to be sent to the window
+ *
+ * @returns {Promise} promise that will resolve to the reply from the window
+ */
+
+export const sendMessageToWindow = (type, payload) =>
```

Comment:
When would we want to do this instead of sending a message to the app / client itself?

Reply:
If you wanna send a message to any client, you would use `sendMessageToClient` and give the client object as a param. If you wanna send a message to `window` you would use `sendMessageToWindow` which uses `sendMessageToClient` internally.